src/thx/schema/macro/SchemaBuilder.hx

```diff
@@ -0,0 +1,264 @@
+  // here are passed things like Option<Array<A>> or Either<Option<String>> where A is a type paramter of the container schema
+  public static function lookupSchema(schemaType: BoundSchemaType, typeSchemas: Map<String, Expr>): Expr {
+    return switch schemaType.type {
+      case LocalParam(param):
+        var name = TypeBuilder.variableNameFromTypeParameter(param);
+        return macro (() -> $i{name});
+      case QualifiedType(_):
+        var type = schemaType.toString();
+        if(typeSchemas.exists(type)) {
+          typeSchemas.get(type);
+        } else {
+          var path = TypeBuilder.ensure(schemaType.toUnboundSchemaType(), typeSchemas);
+          macro $p{path};
+        }
+      case AnonObject(obj):
+        var type = schemaType.toString();
+        if(typeSchemas.exists(type)) {
+          typeSchemas.get(type);
+        } else {
+          var path = TypeBuilder.ensure(schemaType.toUnboundSchemaType(obj.params), typeSchemas);
+          macro $p{path};
+        }
+      case TypeDef(type):
+        var stype = schemaType.toString();
+        if(typeSchemas.exists(stype)) {
+          typeSchemas.get(stype);
+        } else {
+          var type = schemaType.toString();
+          if(typeSchemas.exists(type)) {
+            typeSchemas.get(type);
+          } else {
+            var path = TypeBuilder.ensure(schemaType.toUnboundSchemaType(), typeSchemas);
+            macro $p{path};
+          }
+        }
+    };
+  }
+
+  public static function generateSchema(schemaType: BoundSchemaType, typeSchemas: Map<String, Expr>): Expr {
+    return switch [schemaType.toType(), schemaType.type] {
+      case [TInst(_.get() => cls, _),     BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateClassSchema(cls,    qtype, typeSchemas);
+      case [TEnum(_.get() => enm, _),     BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateEnumSchema(enm,     qtype, typeSchemas);
+      case [TAbstract(_.get() => abs, _), BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateAbstractSchema(abs, qtype, typeSchemas);
+      case [TAnonymous(_.get() => anon),  BoundSchemaTypeImpl.AnonObject(obj)]:
+        generateAnonSchema(anon,    obj,   typeSchemas);
+      case [TType(_.get() => def, _),  BoundSchemaTypeImpl.QualifiedType(qtype)]:
+        generateDefTypeSchema(def, qtype,  typeSchemas);
+      case _: fatal('Cannot generate schema for unsupported type ${schemaType.toString()}');
+    }
+  }
+
+  static function generateClassSchema(cls: ClassType, qtype: QualifiedType<BoundSchemaType>, typeSchemas: Map<String, Expr>) {
+    var fields = cls.fields.get().filter(keepVariables);
+    var n = fields.length;
+    return if(n == 0) {
+      var path = qtype.parts();
+      macro thx.schema.SimpleSchema.object(PropsBuilder.Pure(Type.createEmptyInstance($p{path})));
+    } else {
+      // generate constructor function
+      var constructor = generateClassConstructorF(fields.map(classFieldToFunctionArgument), qtype);
```

Comment:
Since in Haxe constructors are used to enforce object invariants, this seems like it might create a bit of a hole in applications where deserialization could construct "illegal" objects leading to undefined behavior. I know it's more work, but what if for classes and abstracts one were to type-infer which constructor to use based upon the types of the schema arguments passed, and then always route through those constructors?